### DISCUSSION TITLE
semicolons in SDL

### DISCUSSION THREAD

**Discussion by studentiks** [geldata/gel#3464] [Ideas]
Good day!

Currently SDL allows missing semicolons in last line of type definition. Is there a chance that semicolons can be avoided completely in future? Thank you!

![image](https://user-images.githubusercontent.com/6019222/153703555-b8f5c10d-5a15-4a72-9a3d-0039b4321aae.png)

BR,

Roman

**Comment by elprans** [comment]
There is a PR that makes them completely optional (#2302), but there are some reservations on whether it's a good idea.

**Reply by amaster507** [reply]
What are the reservations? The discussion alludes to being in a Slack server.

**Reply by elprans** [reply]
IIRC if was to do with how schema files are merged and that lack of requirement for termination might lead to schemas that parse, but are wrong.  @tailhook might be able to explain better.

**Reply by tailhook** [reply]
I think most important issue is error handling. When we have semicolons, we can emit error and recover parser on next semicolon, so multiple errors could be reported to user at the same time. And also a bunch of other errors would be reported at the right place rather then somewhere where next statement starts (even later, because things that begin next SDL item like `property` and `type` are not reserved keywords and can be type or function names).

Another issue is that it makes easier for other tools to work with the SDL. Say code formatter doesn't have to know every detail of syntax and can format based on `;` and `{..}`. Or someone may want to make a tool that extracts just types and annotations without knowing the whole syntax. We already do a similar thing for our migrations (even though, this is not SDL, the task is similar).

Generally, most languages that don't have mandatory semicolons (or other char), are line-based (e.g. Python). We don't want to make newline a statement terminator, it doesn't work for EdgeQL. There are exceptions like JavaScript, where rules of where lines can be wrapped are quite non-obvious. Another exception is Lua, where you can add newline anywhere, but then there were quite a few bugs where lines are joined unexpectedly (I don't mean bugs of the interpreter, I mean bugs in code that require to put semicolon somewhere to make it correct). So we don't want to repeat that experience.
